Enable SrcSet support with two different input formats for Image tag helper

Krishanthaudayakumara opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Apply the label "bug" or "enhancement" as applicable. -->

## Description / Motivation
<!-- Describe your changes in detail -->
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

This PR implements responsive image support for the `ImageTagHelper` by adding `srcset` and `sizes` attribute functionality, enabling developers to output responsive images using Sitecore's built-in media resizing capabilities. 

### Key Changes:

1. **Added `SrcSet` and `Sizes` Properties**: Extended the `ImageTagHelper` with new properties to support responsive image configuration for both `<sc-img>` and `<img>` tags.

2. **Multiple Input Format Support**: The `SrcSet` property accepts two different input formats for maximum flexibility:
   - **Anonymous objects**: `new object[] { new { w = 800 }, new { mw = 400 } }`
   - **Dictionary arrays**: `new Dictionary<string, object> { {"w", 800} }`

3. **Content SDK Parameter Priority**: Implements the same parameter precedence as Content SDK:
   - `w` (width) takes priority over `mw` (max width) for srcset descriptors
   - Extended support for legacy parameters: `w` > `mw` > `width` > `maxWidth`
   - Only entries with valid width parameters are included in srcset
   - Matches Content SDK's `getSrcSet` function behavior exactly

4. **Enhanced Parameter Merging with Preservation**: 
   - Base `imageParams` are merged with individual `srcSet` parameters
   - `srcSet` parameters override `imageParams` for each srcset entry
   - **Critical Fix**: Added `GetSitecoreMediaUriWithPreservation` method to preserve existing query parameters
   - Preserves essential Sitecore parameters (rev, db, la, vs, ts, hash, ttc, tt, etc.)
   - Follows Content SDK's `{ ...imageParams, ...params }` merge pattern

5. **Dual URL Processing Methods**: 
   - **Legacy `GetSitecoreMediaUri`**: Maintains existing behavior (strips query parameters)
   - **New `GetSitecoreMediaUriWithPreservation`**: Preserves critical parameters for responsive images
   - **`GetMediaLinkForSrcSet`**: Specialized method for srcSet URL generation with parameter preservation
   - Ensures backward compatibility while fixing responsive image display issues

### Content SDK Compatibility:

The implementation now produces identical output to the Content SDK's `Image` component:

**Content SDK (React):**
```jsx
<Image 
  field={imageField} 
  imageParams={{quality: 80}} 
  srcSet={[{w: 800}, {mw: 400}]} 
  sizes="(min-width: 768px) 800px, 400px" 
/>
```

**ASP.NET Core SDK (Razor):**
```html
<sc-img asp-for="@Model.ImageField" 
        image-params='new { quality = 80 }'
        src-set='new object[] { new { w = 800 }, new { mw = 400 } }'
        sizes="(min-width: 768px) 800px, 400px" />
```


### Technical Implementation Details:

1. **Parameter Preservation Architecture**:
   - `MergeParameters()`: Merges base and override parameters using reflection and dictionary handling
   - `GetSitecoreMediaUriWithPreservation()`: Preserves existing URL parameters while adding new ones
   - `ConvertToStringDictionary()`: Converts objects to string dictionaries for URL building
   - `GetMediaLinkForSrcSet()`: Specialized extension method for responsive image URL generation

2. **JsonElement Handling**: Enhanced support for JSON deserialization scenarios to handle `JsonElement` types properly

3. **Helper Methods**: Added `AddResponsiveImageAttributes()` overloads to eliminate code duplication across `Process()`, `GenerateImage()`, and editable markup methods

### Usage Examples:

```html
<!-- Basic responsive image -->
<sc-img asp-for="@Model.HeroImage" 
        src-set='new object[] { new { w = 1200 }, new { w = 800 }, new { w = 400 } }'
        sizes="(min-width: 1200px) 1200px, (min-width: 800px) 800px, 400px" />

<!-- With base parameters (preserves existing URL params) -->
<img asp-for="@Model.ProductImage" 
     image-params='new { quality = 80, format = "webp" }'
     src-set='new object[] { new { w = 800, h = 600 }, new { mw = 400, mh = 300 } }'
     sizes="(min-width: 768px) 800px, 400px" />

<!-- Legacy parameter support -->
<sc-img asp-for="@Model.LegacyImage" 
        src-set='new object[] { new { width = 800 }, new { maxWidth = 400 } }'
        sizes="(min-width: 768px) 800px, 400px" />
```

This implementation ensures feature parity with the Content SDK while maintaining the familiar ASP.NET Core TagHelper syntax, full backward compatibility, and proper parameter preservation for functional image display.

## Testing

- [x] The Unit & Integration tests are passing.
- [x] I have added the necessary tests to cover my changes.

## Terms
<!-- Place an X in the [] to check. -->

<!-- The Code of Conduct helps create a safe space for everyone. We require that everyone agrees to it. -->
- [x] I agree to follow this project's Code of Conduct.

Fix #16

Krishanthaudayakumara committed:
Enable SrcSet support with three different input formats and test cases in ImageTagHelperFixture

Krishanthaudayakumara committed:
Remove unnecessary white spaces remained in changes

Krishanthaudayakumara committed:
Add changes to match with content SDK

Krishanthaudayakumara committed:
Alternative approach for parameter preservation in srcSet functionality

Krishanthaudayakumara committed:
Enhanced parameter support - Added legacy width and maxWidth parameter support, Specific exception handling - Only catch JsonException instead of all exceptions, Better ASP.NET Core compatibility - Replaced HttpUtility with QueryHelpers

Krishanthaudayakumara committed:
Replace var keywords with explicit type declarations

Krishanthaudayakumara committed:
feat: Keep URL parameter preservation for srcset and add comprehensive tests

- Keep GetSitecoreMediaUriWithPreservation
- Add 14 new test cases covering srcset parameter merging, validation, and edge cases
- Fix null handling and zero/negative width filtering in ImageTagHelper
- Ensure development environment compatibility with CDN/edge URLs

Krishanthaudayakumara committed:
Remove unnessasary GetMediaLink method overloading

Krishanthaudayakumara committed:
Updated - Enable SrcSet support with Content SDK approach compatibility and comprehensive test cases added

Updated - Enable SrcSet support with Content SDK approach compatibility and comprehensive test cases added

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/Extensions/SitecoreFieldExtensions.cs:
```diff
@@ -63,6 +140,95 @@ private static string GetSitecoreMediaUri(string url, object? imageParams)
+            }
+        }
+
+        // TODO Review hardcoded matching and replacement
+        Match match = MediaUrlPrefixRegex().Match(url);
+        if (match.Success)
+        {
+            url = url.Replace(match.Value, $"/{match.Groups[1]}/jssmedia/", StringComparison.InvariantCulture);
+        }
+
+        return url;
+    }
+
     [GeneratedRegex("/([-~]{1})/media/")]
     private static partial Regex MediaUrlPrefixRegex();
```
Use the approach of already available GetSitecoreMediaUri method

Krishanthaudayakumara committed:
Add comment explanation of GetMediaLinkForSrcSet preservation and using StringComparer.OrdinalIgnoreCase for the dictionary to ensure consistent parameter key handling

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/ImageTagHelper.cs:
```diff
@@ -230,8 +386,61 @@ private HtmlString MergeEditableMarkupWithCustomAttributes(string editableMarkUp
+                continue;
+            }
+
+            // Use GetMediaLinkForSrcSet to preserve existing URL parameters (like ttc, tt, hash, quality, format)
```
Added the explanation of preservation reason

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/Extensions/SitecoreFieldExtensions.cs:
```diff
@@ -29,6 +31,81 @@ public static partial class SitecoreFieldExtensions
+    /// <returns>Merged parameters as dictionary.</returns>
+    private static Dictionary<string, object?> MergeParameters(object? baseParams, object? overrideParams)
+    {
+        Dictionary<string, object?> result = new Dictionary<string, object?>();
```
Implemented the suggestion

robearlam commented on tests/Sitecore.AspNetCore.SDK.RenderingEngine.Tests/TagHelpers/Fields/ImageTagHelperFixture.cs:
```diff
@@ -739,6 +739,456 @@ public void Process_RenderingChromesAreNotNull_ChromesAreOutput(
+        TagHelperOutput tagHelperOutput)
+    {
+        // Arrange
+        Image image = new() { Src = "http://styleguide/-/media/img/sc_logo.png", Alt = "Sitecore Logo" };
```
Why do we need to declare this image, can't we use the `_image` that the other tests use?

robearlam commented on tests/Sitecore.AspNetCore.SDK.RenderingEngine.Tests/TagHelpers/Fields/ImageTagHelperFixture.cs:
```diff
@@ -739,6 +739,456 @@ public void Process_RenderingChromesAreNotNull_ChromesAreOutput(
+
+        // Assert
+        string content = tagHelperOutput.Content.GetContent();
+        content.Should().Contain("srcset=");
+        content.Should().Contain("600w");
+        content.Should().Contain("300w");
+        content.Should().Contain("sizes=\"(min-width: 768px) 600px, 300px\"");
```
Again, full string compare please

Krishanthaudayakumara replied:
I've replaced the Contains() assertions with regex extraction and full string comparison.

robearlam commented on tests/Sitecore.AspNetCore.SDK.RenderingEngine.Tests/TagHelpers/Fields/ImageTagHelperFixture.cs:
```diff
@@ -739,6 +739,456 @@ public void Process_RenderingChromesAreNotNull_ChromesAreOutput(
+        // Assert
+        tagHelperOutput.Attributes.Should().ContainSingle(a => a.Name == "srcset");
+        string srcSetValue = tagHelperOutput.Attributes["srcset"].Value.ToString()!;
+        srcSetValue.Should().Contain("500w");
+        srcSetValue.Should().Contain("250w");
```
Again, full string compare please

Krishanthaudayakumara replied:
Updated to use full string comparison. The test now verifies the exact srcset format

robearlam commented on tests/Sitecore.AspNetCore.SDK.RenderingEngine.Tests/TagHelpers/Fields/ImageTagHelperFixture.cs:
```diff
@@ -739,6 +739,456 @@ public void Process_RenderingChromesAreNotNull_ChromesAreOutput(
+
+        // Assert
+        string content = tagHelperOutput.Content.GetContent();
+        content.Should().Contain("srcset=");
+        content.Should().Contain("800w");
+        content.Should().Contain("400w");
```
Again, full string compare please

Krishanthaudayakumara replied:
I've replaced the Contains() assertions with regex extraction and full string comparison.

robearlam commented on tests/Sitecore.AspNetCore.SDK.RenderingEngine.Tests/TagHelpers/Fields/ImageTagHelperFixture.cs:
```diff
@@ -739,6 +739,456 @@ public void Process_RenderingChromesAreNotNull_ChromesAreOutput(
+
+    [Theory]
+    [AutoNSubstituteData]
+    public void Process_SrcSetWithImageParamsConflict_SrcSetParametersWin(
```
Lets rename this to `Process_SrcSetWithImageParamsConflict_SrcSetParametersArePreferred`

robearlam commented on tests/Sitecore.AspNetCore.SDK.RenderingEngine.Tests/TagHelpers/Fields/ImageTagHelperFixture.cs:
```diff
@@ -739,6 +739,456 @@ public void Process_RenderingChromesAreNotNull_ChromesAreOutput(
+        // Assert
+        tagHelperOutput.Attributes.Should().ContainSingle(a => a.Name == "srcset");
+        string srcSetValue = tagHelperOutput.Attributes["srcset"].Value.ToString()!;
+        srcSetValue.Should().Contain("w=320");
+        srcSetValue.Should().Contain("quality=75");
+        srcSetValue.Should().Contain("format=jpg"); // Inherited from ImageParams
+        srcSetValue.Should().NotContain("w=1000");
+        srcSetValue.Should().NotContain("quality=50");
```
Again, full string compare please

robearlam commented on tests/Sitecore.AspNetCore.SDK.RenderingEngine.Tests/TagHelpers/Fields/ImageTagHelperFixture.cs:
```diff
@@ -739,6 +739,456 @@ public void Process_RenderingChromesAreNotNull_ChromesAreOutput(
+        // Assert
+        tagHelperOutput.Attributes.Should().ContainSingle(a => a.Name == "srcset");
+        string srcSetValue = tagHelperOutput.Attributes["srcset"].Value.ToString()!;
+        srcSetValue.Should().Contain("100w");
+        srcSetValue.Should().Contain("200w");
+        srcSetValue.Should().Contain("300w");
+        srcSetValue.Should().Contain("400w");
```
Again, full string compare please

robearlam commented on tests/Sitecore.AspNetCore.SDK.RenderingEngine.Tests/TagHelpers/Fields/ImageTagHelperFixture.cs:
```diff
@@ -739,6 +739,456 @@ public void Process_RenderingChromesAreNotNull_ChromesAreOutput(
+        // Assert
+        tagHelperOutput.Attributes.Should().ContainSingle(a => a.Name == "srcset");
+        string srcSetValue = tagHelperOutput.Attributes["srcset"].Value.ToString()!;
+        srcSetValue.Should().Contain("320w");
```
Again, full string compare please

robearlam commented on tests/Sitecore.AspNetCore.SDK.RenderingEngine.Tests/TagHelpers/Fields/ImageTagHelperFixture.cs:
```diff
@@ -739,6 +739,456 @@ public void Process_RenderingChromesAreNotNull_ChromesAreOutput(
+        string srcSetValue = tagHelperOutput.Attributes["srcset"].Value.ToString()!;
+
+        // Should preserve existing parameters and merge with new ones
+        srcSetValue.Should().Contain("h=2001");
+        srcSetValue.Should().Contain("iar=0");
+        srcSetValue.Should().Contain("hash=abc123");
+        srcSetValue.Should().Contain("quality=75");
+
+        // New w parameter should override existing w=3000
+        srcSetValue.Should().Contain("w=320");
+        srcSetValue.Should().NotContain("w=3000");
+        srcSetValue.Should().Contain("320w");
```
Again, full string compare please

robearlam commented on tests/Sitecore.AspNetCore.SDK.RenderingEngine.Tests/TagHelpers/Fields/ImageTagHelperFixture.cs:
```diff
@@ -739,6 +739,456 @@ public void Process_RenderingChromesAreNotNull_ChromesAreOutput(
+        string srcSetValue = tagHelperOutput.Attributes["srcset"].Value.ToString()!;
+
+        // Should transform /~/media/ to /~/jssmedia/
+        srcSetValue.Should().Contain("/~/jssmedia/images/test.jpg");
+        srcSetValue.Should().NotContain("/~/media/images/test.jpg");
+        srcSetValue.Should().Contain("320w");
```
Again, full string compare please

robearlam commented on tests/Sitecore.AspNetCore.SDK.RenderingEngine.Tests/TagHelpers/Fields/ImageTagHelperFixture.cs:
```diff
@@ -739,6 +739,456 @@ public void Process_RenderingChromesAreNotNull_ChromesAreOutput(
+        // Assert
+        tagHelperOutput.Attributes.Should().ContainSingle(a => a.Name == "srcset");
+        string srcSetValue = tagHelperOutput.Attributes["srcset"].Value.ToString()!;
+        srcSetValue.Should().Contain("320w");
+        srcSetValue.Should().Contain("480w");
```
Again, full string compare please

Krishanthaudayakumara replied:
Updated to use full string comparison.

robearlam commented on tests/Sitecore.AspNetCore.SDK.RenderingEngine.Tests/TagHelpers/Fields/ImageTagHelperFixture.cs:
```diff
@@ -739,6 +739,456 @@ public void Process_RenderingChromesAreNotNull_ChromesAreOutput(
+        // Assert
+        tagHelperOutput.Attributes.Should().ContainSingle(a => a.Name == "srcset");
+        string srcSetValue = tagHelperOutput.Attributes["srcset"].Value.ToString()!;
+        srcSetValue.Should().Contain("320w");
+        srcSetValue.Should().Contain("quality=75");
+        srcSetValue.Should().Contain("format=webp");
+        srcSetValue.Should().Contain("dpr=2");
+        srcSetValue.Should().Contain("fit=crop");
```
Again, full string compare please

Krishanthaudayakumara replied:
Updated to use full string comparison.

robearlam commented on tests/Sitecore.AspNetCore.SDK.RenderingEngine.Tests/TagHelpers/Fields/ImageTagHelperFixture.cs:
```diff
@@ -739,6 +739,456 @@ public void Process_RenderingChromesAreNotNull_ChromesAreOutput(
+        // Assert
+        tagHelperOutput.Attributes.Should().ContainSingle(a => a.Name == "srcset");
+        string srcSetValue = tagHelperOutput.Attributes["srcset"].Value.ToString()!;
+        srcSetValue.Should().Contain("320w");
+        srcSetValue.Should().Contain("480w");
+        srcSetValue.Should().Contain("quality=75");
+        srcSetValue.Should().Contain("quality=80");
```
Again, full string compare please

Krishanthaudayakumara replied:
Updated to use full string comparison.

robearlam commented on tests/Sitecore.AspNetCore.SDK.RenderingEngine.Tests/TagHelpers/Fields/ImageTagHelperFixture.cs:
```diff
@@ -739,6 +739,456 @@ public void Process_RenderingChromesAreNotNull_ChromesAreOutput(
+        sut.Process(tagHelperContext, tagHelperOutput);
+
+        // Assert
+        tagHelperOutput.Attributes.Should().Contain(a => a.Name == "srcset");
+        tagHelperOutput.Attributes.Should().Contain(a => a.Name == "sizes");
```
Again, full string compare please

Krishanthaudayakumara committed:
refactor the code to eliminate the unnecessary dictionary copying by directly creating the final mergedParams dictionary

Krishanthaudayakumara committed:
extract the duplicated URL prefix replacement logic into a separate method - ApplyJssMediaUrlPrefix

Krishanthaudayakumara committed:
reorder the methods below the properties in ImageTagHelper.cs - SDK.RenderingEngine

Krishanthaudayakumara committed:
updated ParseSrcSet in - ImageTagHelper.cs code to use the [srcSetValue] return

Krishanthaudayakumara committed:
Remove unnessasary comments in ImageTagHelper and refactor ParseSrcSet to ParseJsonSrcSet, move the type checks to the calling method and simplifying ParseJsonSrcSet

Krishanthaudayakumara committed:
Remove unnessasary comments in ImageTagHelper and refactor ParseSrcSet to ParseJsonSrcSet, move the type checks to the calling method and simplifying ParseJsonSrcSet

Krishanthaudayakumara committed:
Throw exception in invalid JSON passing in ParseJsonSrcSet, updated the test cases to validate the fix

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/ImageTagHelper.cs:
```diff
@@ -29,8 +30,118 @@ public class ImageTagHelper(IEditableChromeRenderer chromeRenderer) : TagHelper
+            try
+            {
+                Dictionary<string, object>[]? parsed = System.Text.Json.JsonSerializer.Deserialize<Dictionary<string, object>[]>(jsonString);
+                return parsed?.Cast<object>().ToArray();
```
Thanks for the suggestion Rob, 
I tried your suggestion to simplify the code, but it caused test failures. When deserializing to object[], System.Text.Json creates JsonElement objects instead of dictionaries, which breaks GetWidthDescriptor() that specifically checks for Dictionary<string, object>.

The explicit deserialization to Dictionary<string, object>[] ensures our JSON objects have the right structure for the width extraction logic. I've added comments to clarify this requirement.

Krishanthaudayakumara committed:
Add clarification comment for the use of Dictionary<string, object>[]

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/ImageTagHelper.cs:
```diff
@@ -29,8 +30,118 @@ public class ImageTagHelper(IEditableChromeRenderer chromeRenderer) : TagHelper
+            catch (System.Text.Json.JsonException)
+            {
+                // If JSON parsing fails, return null to skip srcset generation
+                return null;
```
Thanks Rob, You're absolutely right about not swallowing exceptions. I've addressed this and modified the implementation by throwing an InvalidOperationException.

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/ImageTagHelper.cs:
```diff
@@ -230,8 +386,61 @@ private HtmlString MergeEditableMarkupWithCustomAttributes(string editableMarkUp
+
+    private string GenerateSrcSetAttribute(ImageField imageField)
+    {
+        object[]? parsedSrcSet = ParseSrcSet(SrcSet);
```
I've added the fix. The GenerateSrcSetAttribute method now includes the type check before calling ParseJsonSrcSet.

Krishanthaudayakumara committed:
optimize this by using LINQ's Where() - in GenerateSrcSetAttribute

Krishanthaudayakumara committed:
Remove Unnecessary comment

Krishanthaudayakumara committed:
Add readable approach in priority checking in GetWidthDescriptor method

Krishanthaudayakumara committed:
Remove unnecessary comment

Krishanthaudayakumara committed:
refactor the GetWidthDescriptor method to use a helper method and the null-coalescing pattern for better readability

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/ImageTagHelper.cs:
```diff
@@ -29,8 +30,118 @@ public class ImageTagHelper(IEditableChromeRenderer chromeRenderer) : TagHelper
+                    }
+                }
+            }
+        }
```
Thank you Rob, I've implemented your suggestion with updates for better readability

Krishanthaudayakumara committed:
refactor: improve ImageTagHelper test assertions with precise string comparisons

- Replace Contains() assertions with exact string matching for srcset validation
- Use full URL verification including preserved query parameters and HTML encoding
- Rename Process_SrcSetWithImageParamsConflict_SrcSetParametersWin to Process_SrcSetWithImageParamsConflict_SrcSetParametersArePreferred
- Consolidate image object usage by replacing custom Image instances with existing _image field

Krishanthaudayakumara commented on tests/Sitecore.AspNetCore.SDK.RenderingEngine.Tests/TagHelpers/Fields/ImageTagHelperFixture.cs:
```diff
@@ -739,6 +739,456 @@ public void Process_RenderingChromesAreNotNull_ChromesAreOutput(
+
+        // Assert - <sc-img> generates HTML content, not attributes
+        string content = tagHelperOutput.Content.GetContent();
+        content.Should().Contain("srcset=");
+        content.Should().Contain("400w");
+        content.Should().Contain("200w");
```
I've updated the test to use precise string matching. The test now extracts the exact srcset attribute value using regex and verifies the complete URLs with preserved query parameters and width descriptors.

Krishanthaudayakumara commented on tests/Sitecore.AspNetCore.SDK.RenderingEngine.Tests/TagHelpers/Fields/ImageTagHelperFixture.cs:
```diff
@@ -739,6 +739,456 @@ public void Process_RenderingChromesAreNotNull_ChromesAreOutput(
+        tagHelperOutput.Attributes.Should().ContainSingle(a => a.Name == "srcset");
+        string srcSetValue = tagHelperOutput.Attributes["srcset"].Value.ToString()!;
+
+        // Should only contain the entry with mw parameter
+        srcSetValue.Should().Contain("250w");
```
Updated to use full string comparison.

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/Extensions/SitecoreFieldExtensions.cs:
```diff
@@ -63,6 +140,95 @@ private static string GetSitecoreMediaUri(string url, object? imageParams)
+        foreach (KeyValuePair<string, string> kvp in existingParams)
+        {
+            mergedParams[kvp.Key] = kvp.Value;
+        }
```
Thanks for the suggestion Rob, I've optimized the parameter merging by renaming the existing existingParams dictionary to mergedParams - https://github.com/Sitecore/ASP.NET-Core-SDK/pull/46/commits/10c272af8a7455cb0e52c99258bf628a317ad9a6

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/ImageTagHelper.cs:
```diff
@@ -29,8 +30,118 @@ public class ImageTagHelper(IEditableChromeRenderer chromeRenderer) : TagHelper
+        // Single object - wrap in array
+        return new[] { srcSetValue };
+    }
+
```
I've moved all the static helper methods (ParseJsonSrcSet, GetWidthDescriptor, GetWidthFromParameters, etc.) to appear after the properties in the [ImageTagHelper] class, which resolves the compiler warning about method ordering

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/ImageTagHelper.cs:
```diff
@@ -98,6 +222,22 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
+                    }
+                }
+
+                // Add sizes if provided
```
I've removed the unnecessary comment

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/ImageTagHelper.cs:
```diff
@@ -98,6 +222,22 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
             {
                 output.Attributes.Add(ScrAttribute, field.GetMediaLink(ImageParams));
 
+                // Add srcset if configured
```
I've removed the unnecessary comment

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/ImageTagHelper.cs:
```diff
@@ -152,6 +292,22 @@ private TagBuilder GenerateImage(ImageField imageField, TagHelperOutput output)
+                }
+            }
+
+            // Add sizes if provided
```
I've removed the unnecessary comment

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/ImageTagHelper.cs:
```diff
@@ -230,8 +386,61 @@ private HtmlString MergeEditableMarkupWithCustomAttributes(string editableMarkUp
 
             imageNode.SetAttributeValue(ScrAttribute, imageField.GetMediaLink(ImageParams));
+
+            // Add srcset for editable content
```
I've removed the unnecessary comment

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/ImageTagHelper.cs:
```diff
@@ -230,8 +386,61 @@ private HtmlString MergeEditableMarkupWithCustomAttributes(string editableMarkUp
+                }
+            }
+
+            // Add sizes for editable content
```
I've removed the unnecessary comment

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/ImageTagHelper.cs:
```diff
@@ -29,8 +30,118 @@ public class ImageTagHelper(IEditableChromeRenderer chromeRenderer) : TagHelper
+
+        // If it's a JSON string, parse it
+        if (srcSetValue is string jsonString)
+        {
```
I've updated the method to throw an [InvalidOperationException](vscode-file://vscode-app/c:/Users/KrishanthaKumara/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) with a descriptive message when JSON parsing fails - https://github.com/Sitecore/ASP.NET-Core-SDK/pull/46/commits/ac23fcb2c986339ecd28fbbd991cc8aa87a48666

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/ImageTagHelper.cs:
```diff
@@ -29,8 +30,118 @@ public class ImageTagHelper(IEditableChromeRenderer chromeRenderer) : TagHelper
+
+        string? width = null;
+
+        // Handle Dictionary<string, object>
```
I've removed the unnecessary comment

Krishanthaudayakumara committed:
remove the intermediate variable and casting in ParseJsonSrcSet

Krishanthaudayakumara committed:
Remove some unnecessary comments used in ImageTagHelper

robearlam commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/Extensions/SitecoreFieldExtensions.cs:
```diff
@@ -29,6 +31,81 @@ public static partial class SitecoreFieldExtensions
+    /// <param name="baseParams">Base parameters.</param>
+    /// <param name="overrideParams">Override parameters that take precedence.</param>
+    /// <returns>Merged parameters as dictionary.</returns>
+    private static Dictionary<string, object?> MergeParameters(object? baseParams, object? overrideParams)
```
Can we rename these properties to be `imageParams` and `scrSetParams` to be more explicit

Krishanthaudayakumara replied:
Thanks Rob, I renamed `baseParams` to `imageParams` to better reflect base image parameters and `overrideParams` to `srcSetParams` to indicate srcset specific parameters

robearlam commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/Extensions/SitecoreFieldExtensions.cs:
```diff
@@ -29,6 +31,81 @@ public static partial class SitecoreFieldExtensions
+        // Add base parameters first
+        if (baseParams != null)
+        {
+            if (baseParams is Dictionary<string, object> baseDict)
+            {
+                foreach (KeyValuePair<string, object> kvp in baseDict)
+                {
+                    result[kvp.Key] = kvp.Value;
+                }
+            }
+            else
+            {
+                PropertyInfo[] baseProps = baseParams.GetType().GetProperties();
+                foreach (PropertyInfo prop in baseProps)
+                {
+                    result[prop.Name] = prop.GetValue(baseParams);
+                }
+            }
```
This is duplicated below for the scrSetparams, lets move the logic into a method to reduce the duplication and make the code more readable.

Krishanthaudayakumara replied:
Thanks Rob, I've extracted the duplicated parameter processing logic into a new `AddParametersToResult` helper method. This eliminates the duplication in both the `imageParams` and `srcSetParams` processing within `MergeParameters`.

robearlam commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/Extensions/SitecoreFieldExtensions.cs:
```diff
@@ -29,6 +31,81 @@ public static partial class SitecoreFieldExtensions
+        // Override with srcSet parameters
+        if (overrideParams != null)
+        {
+            if (overrideParams is Dictionary<string, object> overrideDict)
+            {
+                foreach (KeyValuePair<string, object> kvp in overrideDict)
+                {
+                    result[kvp.Key] = kvp.Value;
+                }
+            }
+            else
+            {
+                PropertyInfo[] overrideProps = overrideParams.GetType().GetProperties();
+                foreach (PropertyInfo prop in overrideProps)
+                {
+                    result[prop.Name] = prop.GetValue(overrideParams);
+                }
+            }
```
As mentioned above, combine with code above in a method.

robearlam commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/Extensions/SitecoreFieldExtensions.cs:
```diff
@@ -53,6 +130,89 @@ private static string GetSitecoreMediaUri(string url, object? imageParams)
+        // Add new parameters (these will override existing ones)
+        if (parameters != null)
+        {
+            if (parameters is Dictionary<string, object> paramDict)
+            {
+                foreach (KeyValuePair<string, object> kvp in paramDict)
+                {
+                    mergedParams[kvp.Key] = kvp.Value;
+                }
+            }
+            else
+            {
+                PropertyInfo[] properties = parameters.GetType().GetProperties();
+                foreach (PropertyInfo prop in properties)
+                {
+                    object? value = prop.GetValue(parameters);
+                    if (value != null)
+                    {
+                        mergedParams[prop.Name] = value;
+                    }
+                }
+            }
```
Is this also a case where we could use the method I called out above?

robearlam commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/Extensions/SitecoreFieldExtensions.cs:
```diff
@@ -53,6 +130,89 @@ private static string GetSitecoreMediaUri(string url, object? imageParams)
+
+        // Parse existing query parameters and build merged parameters dictionary
+        Dictionary<string, object?> mergedParams = new Dictionary<string, object?>(StringComparer.OrdinalIgnoreCase);
+        if (url.Contains('?'))
+        {
+            string[] parts = url.Split('?', 2);
+            url = parts[0];
+            string queryString = parts[1];
+
+            string[] paramPairs = queryString.Split('&');
+            foreach (string paramPair in paramPairs)
+            {
+                string[] keyValue = paramPair.Split('=', 2);
+                if (keyValue.Length == 2)
+                {
+                    string key = HttpUtility.UrlDecode(keyValue[0]);
+                    string value = HttpUtility.UrlDecode(keyValue[1]);
+                    mergedParams[key] = value;
+                }
+            }
+        }
```
Move into method called `ParseURLParams`

Krishanthaudayakumara replied:
Sure Rob,  I've extracted the URL parameter parsing logic into a separate `ParseUrlParams` method.

Krishanthaudayakumara replied:
Thanks Rob, I've now moved the complete URL parsing logic into the `ParseUrlParams` method. 

The updated calling code,

```csharp
Dictionary<string, object?> mergedParams = new Dictionary<string, object?>(StringComparer.OrdinalIgnoreCase);
url = ParseUrlParams(url, mergedParams);
AddParametersToResult(mergedParams, parameters, skipNullValues: true);
```

https://github.com/Sitecore/ASP.NET-Core-SDK/pull/46/commits/1343dc7440247d09b5909afe4d5a79d4979c673d

robearlam commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/ImageTagHelper.cs:
```diff
@@ -141,6 +172,79 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
+            try
+            {
+                // We need to use Dictionary<string, object>[] to ensure proper deserialization of JSON objects into dictionaries that our GetWidthDescriptor method can handle
+                return System.Text.Json.JsonSerializer.Deserialize<Dictionary<string, object>[]>(jsonString, JsonLayoutServiceSerializer.GetDefaultSerializerOptions());
```
Introduce a using to remove need to use the Fully Qualified name

Krishanthaudayakumara replied:
Sure Rob, I've added `using System.Text.Json;` to the imports and replaced the fully qualified `System.Text.Json.JsonSerializer` with just `JsonSerializer` in the ParseJsonSrcSet method.

Krishanthaudayakumara committed:
refactor: rename MergeParameters method parameters for clarity

- Rename aseParams to imageParams to better reflect base image parameters
- Rename overrideParams to srcSetParams to indicate srcset-specific parameters

Krishanthaudayakumara committed:
refactor: extract duplicate parameter processing logic into reusable method

- Create AddParametersToResult helper method to handle both Dictionary and reflection-based parameter processing
- Add skipNullValues parameter to handle different null value handling requirements
- Refactor MergeParameters to use the new helper method, eliminating code duplication
- Update GetSitecoreMediaUriWithPreservation to use the helper method with skipNullValues: true

Krishanthaudayakumara committed:
refactor: extract URL parameter parsing into separate ParseUrlParams method - ParseUrlParams

Krishanthaudayakumara committed:
refactor: add using System.Text.Json to remove fully qualified name -  Add using System.Text.Json directive to imports

Krishanthaudayakumara committed:
refactor: move complete URL parsing logic into ParseUrlParams method

Krishanthaudayakumara committed:
refactor: move mergedParams declaration to ParseUrlParams call for cleaner code

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/Extensions/SitecoreFieldExtensions.cs:
```diff
@@ -53,6 +133,83 @@ private static string GetSitecoreMediaUri(string url, object? imageParams)
+
+        // Parse existing query parameters and build merged parameters dictionary
+        Dictionary<string, object?> mergedParams = new Dictionary<string, object?>(StringComparer.OrdinalIgnoreCase);
+        url = ParseUrlParams(url, mergedParams);
```
Sure Rob, I've moved the variable declaration down so both `mergedParams ` and `url ` are initialized together.

sc-ivanlieckens commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/Extensions/SitecoreFieldExtensions.cs:
```diff
@@ -29,6 +31,84 @@ public static partial class SitecoreFieldExtensions
+        }
+        else
+        {
+            PropertyInfo[] properties = parameters.GetType().GetProperties();
```
Why not use a RouteValueDictionary to do the parsing? (See example on line 129)

sc-ivanlieckens commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/Extensions/SitecoreFieldExtensions.cs:
```diff
@@ -53,6 +133,81 @@ private static string GetSitecoreMediaUri(string url, object? imageParams)
+
+        // Parse existing query parameters and build merged parameters dictionary
+        Dictionary<string, object?> mergedParams = new Dictionary<string, object?>(StringComparer.OrdinalIgnoreCase);
+        string url = ParseUrlParams(urlStr, mergedParams);
```
There's also QueryHelpers.ParseQuery if that may be simpler? Also Uri can be a class that gives you strong parsing and performance taking into account quirks and standards.

sc-ivanlieckens commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/ImageTagHelper.cs:
```diff
@@ -141,6 +173,79 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
+        if (parameters is Dictionary<string, object> dictionary)
+        {
+            // Priority: w > mw > width > maxWidth (matching Content SDK behavior + legacy support)
+            if (dictionary.ContainsKey("w"))
```
https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1854

sc-ivanlieckens commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/ImageTagHelper.cs:
```diff
@@ -141,6 +173,79 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
+            try
+            {
+                // We need to use Dictionary<string, object>[] to ensure proper deserialization of JSON objects into dictionaries that our GetWidthDescriptor method can handle
+                return JsonSerializer.Deserialize<Dictionary<string, object>[]>(jsonString, JsonLayoutServiceSerializer.GetDefaultSerializerOptions());
```
Why are we doing this? Serializers outside of binding?

Krishanthaudayakumara replied:
Removed JSON string support for `SrcSet` and this would no longer available.

- Removed support for passing `SrcSet` as a JSON string to `ImageTagHelper`, only object arrays (anonymous objects or dictionaries) are now supported.
- Deleted `ParseJsonSrcSet` and all related JSON parsing logic from `ImageTagHelper`.
- Removed or updated test cases that expected JSON string support for `SrcSet`, including tests for invalid JSON and JSON string srcset generation.

https://github.com/Krishanthaudayakumara/ASP.NET-Core-SDK/pull/14/commits/61610684eaf825edf9e20340a64575596c4243e0

sc-ivanlieckens commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/ImageTagHelper.cs:
```diff
@@ -141,6 +173,79 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
+        return prop?.GetValue(parameters)?.ToString();
+    }
+
+    private static string? GetWidthDescriptor(object? parameters)
```
Is there a better method to read from parameters like using PropertyHelper or HtmlHelper?

Krishanthaudayakumara replied:
Thank you for your input Ivan, 

In this case, I used reflection to read property values from the parameters object, since I couldn't find a public alternative like `PropertyHelper` (which seems to be internal) or anything similar in the framework. If you have a preferred approach or know of a better way to handle this, I’d really appreciate your suggestion and would be happy to update the code accordingly.

robearlam replied:
Could we parse this using the `Microsoft.AspNetCore.Mvc.ViewFeatures.HtmlHelper`?

I think the parsing in there should be able to handle both dictionary and anonymous object use cases in one, meaning we can simplify the method to something like

```
IDictionary<string, object> dictionary = HtmlHelper.AnonymousObjectToHtmlAttributes(parameters);

string? width = null;

// Priority: w > mw > width > maxWidth (matching Content SDK behavior + legacy support)
if (dictionary.TryGetValue("w", out var wValue))
{
    width = wValue.ToString();
}
else if (dictionary.TryGetValue("mw", out var mwValue))
{
    width = mwValue.ToString();
}
else if (dictionary.TryGetValue("width", out var widthValue))
{
    width = widthValue.ToString();
}
else if (dictionary.TryGetValue("maxWidth", out var maxWidthValue))
{
    width = maxWidthValue.ToString();
}

if (width != null && int.TryParse(width, out int widthValueInt) && widthValueInt <= 0)
{
    return null;
}

return width != null ? $"{width}w" : null;
```

Can you give that a test and make sure that both uses cases are still handle correctly?

sc-ivanlieckens commented on tests/Sitecore.AspNetCore.SDK.RenderingEngine.Tests/TagHelpers/Fields/ImageTagHelperFixture.cs:
```diff
@@ -739,6 +739,462 @@ public void Process_RenderingChromesAreNotNull_ChromesAreOutput(
+        string srcsetValue = srcsetMatch.Groups[1].Value;
+
+        // Verify each entry precisely with preserved query parameters and new width parameters
+        srcsetValue.Should().Contain("http://styleguide/-/jssmedia/styleguide/data/media/img/sc_logo.png?iar=0&amp;hash=F313AD90AE547CAB09277E42509E289B&amp;w=400 400w");
```
&amp; ?

Krishanthaudayakumara replied:
The `&amp;` is expected here because the test checks the HTML output, where ampersands(&) in attribute values  encoded as `&amp;`. This ensures the generated srcset is valid HTML and preserves all query parameters. Let me know if I should update anything.

sc-ivanlieckens commented on tests/Sitecore.AspNetCore.SDK.RenderingEngine.Tests/TagHelpers/Fields/ImageTagHelperFixture.cs:
```diff
@@ -739,6 +739,462 @@ public void Process_RenderingChromesAreNotNull_ChromesAreOutput(
+
+        // Verify each entry precisely with preserved query parameters and new width parameters
+        srcsetValue.Should().Contain("http://styleguide/-/jssmedia/styleguide/data/media/img/sc_logo.png?iar=0&amp;hash=F313AD90AE547CAB09277E42509E289B&amp;w=400 400w");
+        srcsetValue.Should().Contain("http://styleguide/-/jssmedia/styleguide/data/media/img/sc_logo.png?iar=0&amp;hash=F313AD90AE547CAB09277E42509E289B&amp;w=200 200w");
```
These look like preview values, did the hashes actually work given they have different "w" values?

robearlam replied:
As discussed, this is matching Content SDK work, it functions against Edge, but not against Preview. Same as with the JS frameworkds

Krishanthaudayakumara committed:
Intergration tests for GeneratesProperSrcSetAttribute and SrcSetAttributeContainsCorrectUrls

Krishanthaudayakumara committed:
Added FourthImage and enhanced integration tests to assert srcset URLs and descriptors for both third and fourth images

Krishanthaudayakumara committed:
update the testcase ImgTagHelper_SrcSetAttributeContainsCorrectUrls to ImgTagHelper_SrcSetAttributeContainsCorrectUrlsAndSizes and clear separation of checking generate srcSet and URL check

Krishanthaudayakumara committed:
Complete check of attribute contents in GenerateProperSrcSetAttribute

Krishanthaudayakumara committed:
Remove ImgTagHelper_SrcSetAttributeContainsCorrectUrlsAndSizes and merge the tests to ImgTagHelper_GeneratesProperSrcSetAttributeWithCorrectUrlsAndSizes

Krishanthaudayakumara committed:
Merge pull request #13 from Krishanthaudayakumara/feat/srcset-support-single-intergration-tests

Intergration tests for GeneratesProperSrcSetAttribute and SrcSetAttributeContainsCorrectUrls

Krishanthaudayakumara committed:
Refactor AddParametersToResult to use RouteValueDictionary for object parameter extraction - SitecoreFieldExtension

Krishanthaudayakumara committed:
Updated ParseUrlParams to use QueryHelpers.ParseQuery for extracting query parameters from URLs

Krishanthaudayakumara committed:
Refactor GetWidthDescriptor to use TryGetValue instead of ContainsKey for CA1854 compliance

Krishanthaudayakumara committed:
Remove JSON string support for SrcSet in ImageTagHelper and update tests

- Removed support for passing SrcSet as a JSON string to ImageTagHelper; only object arrays (anonymous objects or dictionaries) are now supported.
- Deleted ParseJsonSrcSet and all related JSON parsing logic from ImageTagHelper.
- Removed test cases that expected JSON string support for SrcSet - tests for invalid JSON and JSON string srcset generation.

Krishanthaudayakumara committed:
Merge pull request #14 from Krishanthaudayakumara/feat/srcset-support-pr-reviews

Refactor ImageTagHelper Parameter Extraction and Remove JSON SrcSet Support - Code changes of the PR Review
- Refactored AddParametersToResult to use RouteValueDictionary for extracting object properties, improving consistency and maintainability.
- Updated ParseUrlParams to use QueryHelpers.ParseQuery for extracting query parameters from URLs, ensuring more robust and reliable parsing.
- Refactored GetWidthDescriptor to use TryGetValue instead of ContainsKey for CA1854 compliance, improving performance and code clarity.
- Removed JSON string support for SrcSet in ImageTagHelper and updated related tests to reflect this change.

Krishanthaudayakumara committed:
Remove blank line to fix the warning - ComponentWithImage.cs

robearlam commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/Extensions/SitecoreFieldExtensions.cs:
```diff
@@ -53,6 +132,80 @@ private static string GetSitecoreMediaUri(string url, object? imageParams)
+    /// <param name="url">The full URL with potential query parameters.</param>
+    /// <param name="parameters">The dictionary to add parsed parameters to.</param>
+    /// <returns>The URL without query parameters.</returns>
+    private static string ParseUrlParams(string url, Dictionary<string, object?> parameters)
```
Can we update the url param to be of type `Uri` instead as per Ivans comment above.

This will remove the need to do the string parsing below as you can access the required data directly via properties on `Uri`

robearlam commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/Extensions/SitecoreFieldExtensions.cs:
```diff
@@ -53,6 +132,80 @@ private static string GetSitecoreMediaUri(string url, object? imageParams)
+        string cleanUrl = url.Substring(0, queryIndex);
+        string queryString = url.Substring(queryIndex);
+
+        Dictionary<string, Microsoft.Extensions.Primitives.StringValues> parsedQuery = QueryHelpers.ParseQuery(queryString);
+        foreach (KeyValuePair<string, Microsoft.Extensions.Primitives.StringValues> kvp in parsedQuery)
```
Can you introduce a using statement for `Microsoft.Extensions.Primitives` so we don't have to fully qualify `StringValues` in both of these lines?

Krishanthaudayakumara committed:
Refactor GetWidthDescriptor to use HtmlHelper.AnonymousObjectToHtmlAttributes for both anonymous objects and dictionaries

Krishanthaudayakumara committed:
Refactor: Update ParseUrlParams to use Uri type and leverage Uri properties where possible

Krishanthaudayakumara committed:
Refactor: Add using for Microsoft.Extensions.Primitives and use StringValues directly in ParseUrlParams

robearlam commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/Extensions/SitecoreFieldExtensions.cs:
```diff
@@ -53,6 +131,104 @@ private static string GetSitecoreMediaUri(string url, object? imageParams)
+            return string.Empty;
+        }
+
+        string urlWithoutQuery;
+        string? query = null;
+
+        if (uri.IsAbsoluteUri)
+        {
+            urlWithoutQuery = uri.GetLeftPart(UriPartial.Path);
+            query = uri.Query;
+        }
+        else
+        {
+            // For relative URIs, manually split on '?'
+            var original = uri.OriginalString;
+            int idx = original.IndexOf('?');
+            if (idx >= 0)
+            {
+                urlWithoutQuery = original.Substring(0, idx);
+                query = original.Substring(idx);
+            }
+            else
+            {
+                urlWithoutQuery = original;
+            }
+        }
+
+        if (!string.IsNullOrEmpty(query))
+        {
+            var parsedQuery = QueryHelpers.ParseQuery(query);
+            foreach (KeyValuePair<string, StringValues> kvp in parsedQuery)
+            {
+                parameters[kvp.Key] = kvp.Value.Count > 0 ? kvp.Value[0] : null;
+            }
+        }
+
+        return urlWithoutQuery;
```
It feels like we could simplify this massively. Would the following work instead?

```
string url = $"${uri.Scheme}://${uri.Host}";
NameValueCollection queryParams = HttpUtility.ParseQueryString(uri.Query);
foreach (string? param in queryParams.AllKeys)
{
    if (!string.IsNullOrEmpty(param))
    {
        parameters[param] = queryParams[param];
    }
}

return url;
```

Krishanthaudayakumara replied:
Thanks for the suggestion Rob, I tested with this but couple of test cases get failed, since it requires to handle relative URL's as well. So I simplified that in this way - 

```
string original = uri.OriginalString;
int queryIndex = original.IndexOf('?');

if (queryIndex >= 0)
{
    string query = original.Substring(queryIndex);
    var parsedQuery = QueryHelpers.ParseQuery(query);
    foreach (var kvp in parsedQuery)
    {
        parameters[kvp.Key] = kvp.Value.Count > 0 ? kvp.Value[0] : null;
    }
    return original.Substring(0, queryIndex);
}

return original;
```

https://github.com/Sitecore/ASP.NET-Core-SDK/pull/46/commits/b15b4c1a922027c5c2b3cc36a15262706bdeaa78

Krishanthaudayakumara replied:
Update : I've updated the `ParseUrlParams` method to handle absolute and relative URIs separately.
- For absolute URIs, I'm using `Uri` properties (`Scheme`, `Host`, `AbsolutePath`, `Query`) and `HttpUtility.ParseQueryString`.
- For relative URIs, since accessing `Uri.Query` throws `InvalidOperationException`, I'm falling back to string matching and using `QueryHelpers.ParseQuery`.

https://github.com/Sitecore/ASP.NET-Core-SDK/pull/46/commits/ab6e358b9d5c406d1ee6ebd25f87abcea517ac1f

Krishanthaudayakumara committed:
Simplify ParseUrlParams to handle relative URIs without exceptions

Krishanthaudayakumara committed:
refactor: ParseUrlParams - improve URI parsing in SitecoreFieldExtensions to handle absolute and relative URIs separately

Krishanthaudayakumara committed:
Improve URI parsing in SitecoreFieldExtensions to handle relative and absolute URIs separately

- Replace string manipulation with Uri class properties for absolute URIs
- Add proper handling for relative URIs that throw InvalidOperationException when accessing Uri.Query
- Use HttpUtility.ParseQueryString for absolute URIs and QueryHelpers.ParseQuery for relative URIs
- Add explanatory comments for the if/else branching logic
- Simplify ParseUrlParams method while maintaining compatibility with all URI types

Krishanthaudayakumara committed:
Avoid multiple return statements and return null in GetMediaLink

Krishanthaudayakumara committed:
fix  IDE0090 violations in the SitecoreFieldExtensions file by replacing the explicit type declarations with target-typed new expressions

Krishanthaudayakumara committed:
Refactor AddParametersToResult using a switch expression pattern and LINQ to eliminate the early return statement

Krishanthaudayakumara committed:
set blank line after closing brace of following change

Krishanthaudayakumara committed:
remove the redundant else block and consolidate to a single return statement in the ParseUrlParams method

Krishanthaudayakumara committed:
uses the range indexer original[queryIndex..] in parseUrlParams

Krishanthaudayakumara committed:
Remove var - use explicit types instead of var in changes

Krishanthaudayakumara committed:
use original[..queryIndex]; in ParseUrlParams

Krishanthaudayakumara committed:
Remove unnessasary if branch in GenerateSrcSetAttribute - ImageTagHelpers

Krishanthaudayakumara committed:
use collection initializer syntax to create the list in GenerateSrcSetAttribute - ImageTagHelper

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/Extensions/SitecoreFieldExtensions.cs:
```diff
@@ -29,6 +31,83 @@ public static partial class SitecoreFieldExtensions
+
+        if (urlStr == null)
+        {
+            return null;
```
Thanks for the feedbacksRefactored GetMediaLink to use a single return statement - https://github.com/Krishanthaudayakumara/ASP.NET-Core-SDK/pull/26

sc-ivanlieckens replied:
Same adjustment to GetMediaLinkForSrcSet?

Krishanthaudayakumara replied:
Sure Ivan, Did the same for GetMediaLinkForSrcSet here - https://github.com/Krishanthaudayakumara/ASP.NET-Core-SDK/pull/30/files

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/Extensions/SitecoreFieldExtensions.cs:
```diff
@@ -29,6 +31,83 @@ public static partial class SitecoreFieldExtensions
+    {
+        if (parameters == null)
+        {
+            return;
```
I get the point Ivan, So refactored `AddParametersToResult` to use a switch pattern, eliminating the early return and improving readability - https://github.com/Krishanthaudayakumara/ASP.NET-Core-SDK/pull/26

sc-ivanlieckens replied:
And what did you think of the RouteValueDictionary constructor?

Krishanthaudayakumara replied:
I guess the RouteValueDictionary constructor is a handy escape hatch , it accepts IDictionary, POCOs and anonymous objects , normalises keys and copies values so we don’t need lots of special cases.

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/Extensions/SitecoreFieldExtensions.cs:
```diff
@@ -53,6 +132,104 @@ private static string GetSitecoreMediaUri(string url, object? imageParams)
+            {
+                string query = original.Substring(queryIndex);
+                var parsedQuery = QueryHelpers.ParseQuery(query);
+                foreach (var kvp in parsedQuery)
```
Replaced all `var` usages with explicit types throughout the changes - https://github.com/Krishanthaudayakumara/ASP.NET-Core-SDK/pull/26

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/Extensions/SitecoreFieldExtensions.cs:
```diff
@@ -53,6 +132,104 @@ private static string GetSitecoreMediaUri(string url, object? imageParams)
+                    parameters[kvp.Key] = kvp.Value.Count > 0 ? kvp.Value[0] : null;
+                }
+
+                return original.Substring(0, queryIndex);
```
Sure thing. Changed `return original.Substring(0, queryIndex);` to `return original[..queryIndex];` for consistency with modern C# syntax. - https://github.com/Krishanthaudayakumara/ASP.NET-Core-SDK/pull/26

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/ImageTagHelper.cs:
```diff
@@ -141,6 +170,42 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
+
+        // Priority: w > mw > width > maxWidth (matching Content SDK behavior + legacy support)
+        string? width = null;
+        if (dictionary.TryGetValue("w", out var wValue))
```
Replaced all var usages with explicit types throughout the changes - https://github.com/Krishanthaudayakumara/ASP.NET-Core-SDK/pull/26

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/ImageTagHelper.cs:
```diff
@@ -141,6 +170,42 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
+        {
+            width = wValue.ToString();
+        }
+        else if (dictionary.TryGetValue("mw", out var mwValue))
```
Replaced all var usages with explicit types throughout the changes - https://github.com/Krishanthaudayakumara/ASP.NET-Core-SDK/pull/26

Krishanthaudayakumara commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/TagHelpers/Fields/ImageTagHelper.cs:
```diff
@@ -141,6 +170,42 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
+        {
+            width = widthValue.ToString();
+        }
+        else if (dictionary.TryGetValue("maxWidth", out var maxWidthValue))
```
Replaced all var usages with explicit types throughout the changes - https://github.com/Krishanthaudayakumara/ASP.NET-Core-SDK/pull/26

Krishanthaudayakumara committed:
Refactor SrcSet Implementation changes based on PR review - 2025-08-18 - Style, Readability, and Improvements

- Refactored methods to avoid multiple return statements and redundant branches.
- Replaced explicit type declarations with target-typed new expressions (IDE0090).
- Eliminated early returns in favor of switch/case patterns for better readability.
- Updated substring extraction and return statements to use C# range indexers.
- Removed all var usages, replacing them with explicit types.
- Removed unnecessary null checks and used collection initializer syntax for lists.
- Improved overall code clarity, consistency, and compliance with project style guidelines.

sc-ivanlieckens commented on src/Sitecore.AspNetCore.SDK.RenderingEngine/Extensions/SitecoreFieldExtensions.cs:
```diff
@@ -53,6 +127,102 @@ private static string GetSitecoreMediaUri(string url, object? imageParams)
+    /// <param name="uri">The Uri with potential query parameters.</param>
+    /// <param name="parameters">The dictionary to add parsed parameters to.</param>
+    /// <returns>The URL without query parameters.</returns>
+    private static string ParseUrlParams(Uri? uri, Dictionary<string, object?> parameters)
```
Did we squeeze 3 functions into 1? Seeing all these returns I would consider perhaps extracting them as a function on their own and isolating those returns allowing you to do the "branch selection" in this function. As I mentioned earlier in this same file, I dislike multiple return statements as it makes debugging harder.

Krishanthaudayakumara replied:
Make sense Ivan, I refactored URL parsing to single return style with extracting `ParseAbsoluteUriParams` and `ParseRelativeUriParams` as separate methods - https://github.com/Krishanthaudayakumara/ASP.NET-Core-SDK/pull/30/files

Krishanthaudayakumara committed:
Remove multiple return statements - GetMediaLinkForSrcSet

Krishanthaudayakumara committed:
Refactor URL parsing to single-return style
- Extract ParseAbsoluteUriParams and ParseRelativeUriParams; make ParseUrlParams a single-return wrapper.
- Convert ParseRelativeUriParams and GetSitecoreMediaUriWithPreservation to single-return style.
- Preserve existing parsing behavior for absolute and relative URIs.

Krishanthaudayakumara committed:
Keep null semantics and the explicit IsNullOrEmpty guard, and use a single return

Krishanthaudayakumara committed:
Ensure we only drop the path when URI parsing succeedsin GetSitecoreMediaUriWithPreservation

Krishanthaudayakumara committed:
Refactor URL parsing and `GetMediaLinkForSrcSet `to single-return style - PR Review - 09/15

- Extract ParseAbsoluteUriParams and ParseRelativeUriParams; make ParseUrlParams a single-return wrapper.
- Convert ParseRelativeUriParams and GetSitecoreMediaUriWithPreservation to single-return style.
- Preserve existing parsing behavior for absolute and relative URIs.
- Improve debuggability and reduce branching/early return
